File: alarmer/models/history.py
import csv
import pathlib
import os


HISTORY_COLUMN_ID = 'ID'
HISTORY_COLUMN_START_ALARM_TIME = 'START_ALARM_TIME'
HISTORY_COLUMN_STOP_ALARM_TIME = 'STOP_ALARM_TIME'
HISTORY_COLUMN_INTERVAL = 'INTERVAL'


class CsvModel(object):
  """CSVファイルを扱うクラスの親クラス"""
  def __init__(self, csv_file):
    self.csv_file = csv_file
    if not os.path.exists(self.csv_file):
      pathlib.Path(self.csv_file).touch()


class HistoryModel(CsvModel):
  """アラームの履歴を読み書きするクラス"""
  def __init__(self):
    csv_file = self._get_csv_file_name()
    super().__init__(csv_file)
    self.history_data = []

  
  def _get_csv_file_name(self):
    """CSVファイルの名前を設定する
    
    Returns:
      csv_file_name (str): CSVファイルの名前
    """
    csv_file_name = None

    try:
      import settings
      if settings.CSV_FILE_NAME:
        csv_file_name = settings.CSV_FILE_NAME
    except ImportError:
      pass

    if csv_file_name is None:
      csv_file_name = 'history.csv'

    return csv_file_name
  

  def load_history_exclude_header(self):
    """CSVファイルを読み込む（ヘッダー意外）。
        
    Returns:
       self.history_data (list): CSVファイルの内容（ヘッダー意外）
    """
    with open(self.csv_file, mode = 'r') as csv_file:
      reader = csv.DictReader(csv_file)

      for row in reader:
        history_data_elements = [
          row[HISTORY_COLUMN_START_ALARM_TIME],
          row[HISTORY_COLUMN_STOP_ALARM_TIME],
          row[HISTORY_COLUMN_INTERVAL]
        ]
        self.history_data.append(history_data_elements)

    return self.history_data


  def load_history_include_header(self):
    """CSVファイルを読み込む（ヘッダーも）
    
    Returns:
      rows (list): CSVファイルの内容（ヘッダーも）
    """
    with open(self.csv_file, mode = 'r') as csv_file:
      reader = csv.reader(csv_file)
      rows = [row for row in reader]

    for row in rows[0]:
      print(row, end='　')
    print()
    for j in range(1, len(rows)):
      print(rows[j][0], end='')
      print('　　　', rows[j][1], end='')
      print('　　　　', rows[j][2], end='')
      print('　　　', rows[j][3])

    return rows


  def save(self, history_data):
    """CSVファイルに書き込む
    
    Args:
      history_data (list): アラームの履歴
    """
    csv_file_column = [
      HISTORY_COLUMN_ID,
      HISTORY_COLUMN_START_ALARM_TIME,
      HISTORY_COLUMN_STOP_ALARM_TIME,
      HISTORY_COLUMN_INTERVAL
    ]
    with open(self.csv_file, mode='w') as csv_file:
      writer = csv.DictWriter(csv_file, fieldnames=csv_file_column)
      writer.writeheader()
      for id, data in zip(range(1, len(history_data)+1), history_data):
        writer.writerow({
          HISTORY_COLUMN_ID : id,
          HISTORY_COLUMN_START_ALARM_TIME: data[0],
          HISTORY_COLUMN_STOP_ALARM_TIME : data[1],
          HISTORY_COLUMN_INTERVAL : data[2]
        })

File: alarmer/c_alarmer/history.cpp
#include "history.h"
#include "console.h"


#include <fstream>
#include <sstream>
#include <iostream>


CsvModel::CsvModel(std::string csv_file)
{
  this->csv_file = csv_file;
  if (this->csv_file.empty()) {
    this->csv_file = "history.csv";
  }
  std::ifstream infile(this->csv_file.c_str());
  if(!infile){
    std::ofstream outfile(this->csv_file.c_str());
    outfile.close();
  }
}


HistoryModel::HistoryModel(std::string csv_file) : CsvModel(csv_file){}


std::vector<HistoryColumnWithoutId> HistoryModel::load_history_exclude_header()
{
  std::ifstream in_csv_file(csv_file);
  std::string history_data_element;
  std::string cell;
  HistoryColumnWithoutId data = {};

  std::getline(in_csv_file, history_data_element);

  while (std::getline(in_csv_file, history_data_element)) {
    std::istringstream cells(history_data_element);

    std::getline(cells, cell, ',');

    std::getline(cells, cell, ',');
    data.history_column_start_alarm_time = cell;

    std::getline(cells, cell, ',');
    data.history_column_stop_alarm_time = cell;

    std::getline(cells, cell, ',');
    data.history_column_interval = cell;

    history_data.push_back(data);
  }

  in_csv_file.close();
  return history_data;
}


std::vector<HistoryColumn> HistoryModel::load_history_include_header()
{
  std::ifstream in_csv_file(csv_file);
  std::string row;
  std::string cell;
  HistoryColumn data;

  std::vector<HistoryColumn> rows;

  while(std::getline(in_csv_file, row)){
    std::istringstream cells(row);

    std::getline(cells, cell, ',');
    data.history_column_id = cell;

    std::getline(cells, cell, ',');
    data.history_column_start_alarm_time = cell;

    std::getline(cells, cell, ',');
    data.history_column_stop_alarm_time = cell;

    std::getline(cells, cell, ',');
    data.history_column_interval = cell;

    rows.push_back(data);
  }

  std::cout << rows[0].history_column_id<<"　";
  std::cout << rows[0].history_column_start_alarm_time<<"　";
  std::cout << rows[0].history_column_stop_alarm_time<<"　";
  std::cout << rows[0].history_column_interval<<"　";
  std::cout << std::endl;

  for (int i=1; i<rows.size(); i++) {
    std::cout << rows[i].history_column_id;
    std::cout << "　　　　" << rows[i].history_column_start_alarm_time;
    std::cout << "　　　　　　" << rows[i].history_column_stop_alarm_time;
    std::cout << "　　　　　" << rows[i].history_column_interval << std::endl;
  }

  in_csv_file.close();
  return rows;
}


void HistoryModel::save(const std::vector<HistoryColumnWithoutId>& history_data)
{
  std::vector<std::string> csv_file_column = {
    HISTORY_COLUMN_ID,
    HISTORY_COLUMN_START_ALARM_TIME,
    HISTORY_COLUMN_STOP_ALARM_TIME,
    HISTORY_COLUMN_INTERVAL,
  };

  std::ofstream out_csv_file(csv_file);
  out_csv_file << csv_file_column[0] << ","
  << csv_file_column[1] << ","
  << csv_file_column[2] << ","
  << csv_file_column[3] << "\n";
  for (int id =1; id<=history_data.size(); id++) {
    auto& data = history_data[id-1];
    out_csv_file << id << ","
    << data.history_column_start_alarm_time << ","
    << data.history_column_stop_alarm_time << ","
    << data.history_column_interval << "\n";
  }
  out_csv_file.close();
}